sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala

```diff
@@ -2025,9 +2030,74 @@ class Analyzer(override val catalogManager: CatalogManager)
                   name, other.getClass.getCanonicalName)
               }
             }
-          case u @ UnresolvedFunction(funcId, arguments, isDistinct, filter, ignoreNulls) =>
+          case UnresolvedFunction(NonSessionCatalogAndIdentifier(v2Catalog, ident), arguments,
+            isDistinct, filter, _) if v2Catalog.isFunctionCatalog =>
+            val unbound = v2Catalog.asFunctionCatalog.loadFunction(ident)
+
+            val inputType = StructType(arguments.zipWithIndex.map {
+              case (exp, pos) => StructField(s"_$pos", exp.dataType, exp.nullable)
+            })
+
+            val bound = try {
+              unbound.bind(inputType)
+            } catch {
+              case unsupported: UnsupportedOperationException =>
+                failAnalysis(s"Function '${unbound.name}' cannot process input: " +
+                  s"(${arguments.map(_.dataType.simpleString).mkString(", ")}): " +
+                  unsupported.getMessage)
+            }
+
+            bound match {
+              case scalarFunc: ScalarFunction[_] =>
+                if (isDistinct) {
+                  throw QueryCompilationErrors.functionWithUnsupportedSyntaxError(
+                    scalarFunc.name(), "DISTINCT")
+                } else if (filter.isDefined) {
+                  throw QueryCompilationErrors.functionWithUnsupportedSyntaxError(
+                    scalarFunc.name(), "FILTER clause")
+                } else {
+                  // TODO: implement type coercion by looking at input type from the UDF. We may
+                  //  also want to check if the parameter types from the magic method match the
+                  //  input type through `BoundFunction.inputTypes`.
+                  val argClasses = inputType.fields.map(_.dataType)
+                  findMethod(scalarFunc, ScalarFunction.MAGIC_METHOD_NAME, Some(argClasses)) match {
+                    case Some(_) =>
+                      val caller = Literal.create(scalarFunc, ObjectType(scalarFunc.getClass))
+                      Invoke(caller, ScalarFunction.MAGIC_METHOD_NAME, scalarFunc.resultType(),
+                        arguments, returnNullable = scalarFunc.isResultNullable)
+                    case _ =>
+                      // TODO: handle functions defined in Scala too - in Scala, even if a
+                      //  subclass do not override the default method in parent interface defined
+                      //  in Java, the method can still be found from `getDeclaredMethod`.
+                      findMethod(scalarFunc, "produceResult", Some(Seq(inputType))) match {
```

Comment:
I'm a bit confused. `produceResult` is defined in the interface. Why do we need reflection here?